src/Components/Endpoints/test/RazorComponentResultTest.cs

```diff
@@ -368,7 +368,7 @@ public async Task StreamingRendering_CanBeDisabledForSubtree()
     {
         // Arrange
         var testContext = PrepareVaryStreamingScenariosTests();
-        var initialOutputTask = Task.WhenAll(testContext.Renderer.NonStreamingPendingTasks);
+        var initialOutputTask = Task.WhenAll(testContext.Renderer.AllNonStreamingPendingTasks);
```

Comment:
I had to introduce a class for non streaming pending tasks because of these tests. By the time the execution reaches this line in the test, the list was already cleared and `Assert.False(initialOutputTask.IsCompleted);` fails.

Reply:
Could we find a way to accomplish the same thing without allocating an extra list of tasks? Maybe we could find an approach that doesn't involve clearing the `_nonStreamingPendingTasks` list so we can keep the same approach as what we had before.

For example, something like this might work:

```csharp
private async Task WaitForNonStreamingPendingTasks()
{
    while (true)
    {
        // Create a Task that represents the remaining ongoing work for the rendering process
        var pendingWork = Task.WhenAll(_nonStreamingPendingTasks);

        if (pendingWork.IsCompletedSuccessfully)
        {
            // Since all tasks completed synchronously, that means each pending task
            // had already successfully run to completion, and there's no extra work to do.
            break;
        }

        // Wait for streaming operations to complete before checking again.
        await pendingWork;
    }
}
```

Or, a possibly better approach might be to change the visibility of `WaitForNonStreamingPendingTasks()` to `internal` and then have it lazily run a Task based on the current implementation. Something like this:

```csharp
private Task? _waitForNonStreamingPendingTasks;

// ...

internal Task WaitForNonStreamingPendingTasks()
{
    return _waitForNonStreamingPendingTasks ??= Execute();

    async Task Execute()
    {
        while (_nonStreamingPendingTasks.Count > 0)
        {
            // Create a Task that represents the remaining ongoing work for the rendering process
            var pendingWork = Task.WhenAll(_nonStreamingPendingTasks);

            // Clear all pending work.
            _nonStreamingPendingTasks.Clear();

            // new work might be added before we check again as a result of waiting for all
            // the child components to finish executing SetParametersAsync
            await pendingWork;
        }
    }
}
```

That way, you could change the line in this test to:

```csharp
var initialOutputTask = testContext.Renderer.WaitForNonStreamingPendingTasks();
```

What do you think?

Reply:
Actually, that 2nd example could be improved further if you keep `WaitForNonStreamingPendingTasks` private, but then have a public property to access `_waitForNonStreamingPendingTasks` and throw if it's `null`. That way you don't accidentally call `WaitForNonStreamingPendingTasks` from a test before any tasks have been actually added to the list (because then `_waitForNonStreamingPendingTasks` would be stuck in a "completed" state and any future non-streaming pending tasks get ignored).

Reply:
Thanks, Mackinnon. I think it is a great idea! Indeed, I don't need to keep all the non streaming rendering tasks, it is enough to save the WaitForNonStreamingPendingTasks() into a variable and check its completion in these tests.

Note: the class is internal and that is why WaitForNonStreamingPendingTasks() is also internal. I didn't mark it explicitly as internal because there are many other public methods in this class.